### DISCUSSION TITLE
Things are so sluggish after 3 weeks or so

### DISCUSSION THREAD

**Discussion by vanh20** [Security-Onion-Solutions/securityonion#3162] [Unsupported Versions] [answered]
Hello There!

   I have an i5-6600T processor, 16GB 2133MHz SODIMM, and a 1TB TOSHIBA MQ01ABD100 hdd. My box is a M700 Tiny ThinkCentre cpu with usb to ethernet (100MBps) as my management port. All is connected to a 5 port managed switch/SPAN with pfSense fw as the gateway. Fairly old hardware but still working great. I only have this setup for home use, both for my wife's gadgets and mine.
   For more than two weeks, things were sweet after my standalone installation. Things are not so snappy but the response is quite bearable. However, days ago, I noticed the drive's light keeps steady. Shutting it down takes forever, and for "so-status" to yield all OK result takes 21 minutes or so to happen. I also noticed 190K (and growing) when I run "so-redis-count". Drive, according to grafana /nsm, is at 86%. Memory is also full with just around 500MB+ free.
    I'm a newbie into SO that I am stuck. I spent days already trying to figure this out, but I really do not know why things turned out this way. Is there anything that I missed in reading the manual? Is there a configuration I failed to do? Will things be better if I make my RAM 32GB? Will a 1TB TLC SDD help or will this use case simply kill the drive sooner than expected? I am really lost here. 
   Thanks for your time, and more power to you guys! :)

![screenshot-10 0 100 11-2021 02 27-00_14_40](https://user-images.githubusercontent.com/79697236/109325648-2453e100-7891-11eb-8b54-2d7bb4f00310.png)
![screenshot-10 0 100 11-2021 02 27-00_15_33](https://user-images.githubusercontent.com/79697236/109325669-2c138580-7891-11eb-862d-7cf92c0c06b2.png)
![screenshot-10 0 100 11-2021 02 27-00_16_09](https://user-images.githubusercontent.com/79697236/109325692-3170d000-7891-11eb-90dd-1f52df7a94c9.png)

**Comment by eduncan911** [comment]
I haven't setup SO yet.  However, typically timeseries and logging setups like what SO does on the backend typically need specific data retention policies that will clean up data after X days and/or Y size.

**Reply by vanh20** [reply]
Hello Ed!
   Thanks for the reply! :) I'm also thinking along those lines. I was actually worried that what I was triaging would disappear after just a few days since I got so busy with other things for a while. But to my surprise, things got retained for more than two weeks already. Security Onion seem pretty much configured right out of the box. I'm actually impressed by it so much that I didn't pay any attention to tweaking/tuning it.

   It says here (https://docs.securityonion.net/en/2.3/logstash.html?highlight=rollover), however, that logs are deleted in 7 days. Also, viewing /opt/so/conf/logstash/etc/log4j2.properties tells me there is a "DefaultRolloverStrategy" at work. Which "appender.rolling.strategy.action.condition.nested_condition.age = 7D" confirms to be 7 days, I presume. I have no need to make it longer.

  AS I have already spent many 12 to 16 hour whole days figuring this out, I think I would just reinstall since I do not have much time managing SO itself. I will leave that to the developers. But for what it's worth, I wish there's some best practice table/guide we can follow IF tweaking it and making it run really falls on the end user's responsibility. If I missed reading some docs, I would be more than happy to be advised.

   I'm a great believer of SO. If 2.3 makes it so difficult for me, I would have to fall back to 16.04. In the mean time, I wish there is a way, again happy to be advised here, to carry over the collected logs into the new install. I also encountered another problem when changing the IP address of my SO standalone installation, but I would post it in another thread instead.

   Thanks everyone for your time! :)

**Reply by eduncan911** [reply]
Nice. And I agree, there should be a tuning guide or something that exposes the params that can be changed.  And some of these workflows depend on another app's settings.  Knowing the tech stack of SO, where I've setup and optimized most of the backends individually, it would have to be a number of apps configured at once.  These days, CICD best practices involves the use of an automation tool with a templating engine that can write multiple configs from the same values.  SaltStack, Chef Automate, Ansible, Puppet, etc.

Which led me here; https://docs.securityonion.net/en/2.3/salt.html#configuration

There are other things to clean like Influx and Redis, which can fill up a disk.  There may be more data stores.

OT: I like to poke around a project's community and checkout the overall feedback before I commit. This SO project is one of the best examples of CICD, right down to the automatic builds.

**Reply by vanh20** [reply]
Now, how cool is that! :)

   It may take a while, but now I see the rabbit trail! :D
 (/opt/so/saltstack/local/pillar/global.sls etc.).

Thanks Man! :)

**Comment by pcs4fun** [comment, marked as answer]
Hi,

Sorry to hear about your slow-down.

My setup seems roughly similar to yours and I haven’t had a slow-down issue (yet). Currently I have 6 weeks of data.

Its a home setup, using a Dell T30 with an E3-1225 v5 CPU, 16GB DDR4-2400 RAM, and a 1TB Samsung 860 EVO SSD on Ubuntu 18.04.5 server.

One key difference in systems is HDD vs SSD. 

When comparing your Grafana plots against mine I found the following differences.

- Higher CPU utilization: 75% vs 22%.
- Higher number of ES Documents and much larger ES Store Size.
    My ES document number is 60% of yours, and the Store Size is 13% of yours
    after accumulating data for 6 weeks. So it looks like your storing much more data.
- No Zeek statistics. (Maybe you are not using Zeek?)
- **Much higher IO Wait times: 1.2 min for yours, 2 sec for mine**. (HHD vs SSD?)

During the install I only selected Grafana and Wazuh in an attempt to minimize memory usage.

Might be worth trying an SSD.

With only 16GB of RAM I have about 1GB of swap usage. Comparing SSD LBA writes using 'smartctl' against Samsung's LBA lifetime spec predicts a 15 year life span for the SSD even with the swap usage.

Best of luck!!!

**Reply by vanh20** [reply]
Hello!
   Thanks for taking time to comment and interpret my grafana graphs. Also for sharing your SSD drive usage and lifetime estimate, and for your thoughts in general. They are extremely helpful and I appreciate this a lot! :)

   I will try adding another 16GB ram and replacing the drive today with an SSD since I'm also having system heating issue. Also, I did install everything... (maybe the cause?) I'm following this pipe as well.

   Out of curiosity, I ran the following command just to see how much swap is being used.

> $ swapon -s
> Filename				        Type		Size	Used	Priority
> /dev/dm-1                              	partition	        8388604	4353432	-2

   Clearly, my system is indeed crawling due to lack of memory! This bid me to take an NMON snapshot just to see the overall performance of my system and my suspicion is confirmed! It's right down below for reference.

   The recommendation given [here](https://docs.securityonion.net/en/2.3/hardware.html?highlight=SSD) (Elastic Stack section) is note worthy as far as performance is concerned. I thought I could get by without it and be okay. :D

   I'm not exactly sure why Zeek is not working (probably due to lack of memory?). I can remember on my first install, it was working just fine. Also, so-status yielded an "OK" remark ATM. I will try to conclude this in a few days after I have tried the upgrades.

Thank you all for your time! :)

![SO node metrics](https://user-images.githubusercontent.com/79697236/109406240-835d4700-79b2-11eb-9bb2-889cff81f1de.png)

**Reply by TOoSmOotH** [reply]
I would make sure so-sensor-clean is not spawning multiple processes. This can slow things down considerably. There is a fix coming in 2.3.30.

**Reply by vanh20** [reply]
Hey Mike, Thanks for that! I love this part (/root/SecurityOnion/salt/common/tools) most of all. :)

   I did upgrade my ram to 32GB. It was sweet for a while. All went snappy except for the so-status result which still took about 21 to 22 minutes since uptime before all are fully loaded. My redis backlog started to come down and stayed around the area of 150k till this moment. Last night, SO started using swap a few hours later very slowly. This morning, its now 2GB++, and right now it's 5.8GB++ already and still climbing.

`    After I ran so-sensor-clean, so-status gave a very responsive reply, just matters of seconds and not minutes unlike last night. Also I noticed my PCAP retention also went from 2++ weeks to matters of 6 days and even dropping still until it reached 5.53 days and started climbing again. Maybe that's what the command does?

   Top also didn't yield multiple instances of "so-sensor-clean" ATM. But not sure it was so before I ran the command.
    
   Anyway, I decided to do a fresh install on an SSD, as I believe I already have way too much free memory(?) and SO will continue to use swap space anyway(?). And it's only a matter of time before things become sluggish again(?). It's the managing of read and write delays that have to be dealt with IMHO. Will also adjust redis to 8GB as recommended in the documentation. I'm including my final screen shots for reference.

 This is a great project.

Thanks everyone for your time! :)

![screenshot-10 0 100 11-2021 03 01-10_57_41](https://user-images.githubusercontent.com/79697236/109451255-03091580-7a88-11eb-855f-972978a29f94.png)
![screenshot-10 0 100 11-2021 03 01-11_06_28](https://user-images.githubusercontent.com/79697236/109451293-16b47c00-7a88-11eb-802a-f2b4e2a912c1.png)
![screenshot-10 0 100 11-2021 03 01-11_07_07](https://user-images.githubusercontent.com/79697236/109451306-1caa5d00-7a88-11eb-8b28-7435c49bbb66.png)
![Screenshot from 2021-03-01 11-47-45](https://user-images.githubusercontent.com/79697236/109451316-2207a780-7a88-11eb-8193-882f38ac79f9.png)

**Comment by Acewiza** [comment]
Probably on the right track looking at SSD storage.  I haven't done any real systematic observations or testing myself, but SO "seems" to me to be rather disk-intensive, by default.  My old i5 also with 16Gb RAM, seems to do real well, but a home network with a web server isn't going to generate alot of traffic under most circumstances.  It uses a 4-disk/3.5Tb SSD raid0 for NSM sitting at 57% full after almost 3 months now.  The only issue I ever have with it is some service erring out and unable to restart every couple weeks.

It's all default, with all services selected for activation.

**Reply by vanh20** [reply]
Hi There!

   Thanks for weighing in! :)  Yes, SO indeed is very disk intensive, especially when all services are enabled. This makes the recommendation for SSDs even stronger. But one needs to over-provision and make sure of his SSD's metrics like you did. :)

Again, Thanks! :)
